src/config.rs

```diff
@@ -9,20 +9,90 @@
 // except according to those terms.
 
 use std::path::Path;
+use std::fmt::Debug;
+
+use serde::de::{Deserialize, Deserializer};
 
 use rustfmt::config::Config as RustfmtConfig;
 use rustfmt::config::WriteMode;
 
 const DEFAULT_WAIT_TO_BUILD: u64 = 500;
 
+/// Some values in the config can be inferred without an explicit value set by
+/// the user. There are no guarantees which values will or will not be passed
+/// to the server, so we treat deserialized values effectively as `Option<T>`
+/// and use `None` to mark the values as unspecified, otherwise we always use
+/// `Specified` variant for the deserialized values. For user-provided `None`
+/// values, they must be `Inferred` prior to usage (and can be further
+/// `Specified` by the user).
+#[derive(Clone, Debug, Serialize)]
+pub enum Inferrable<T> {
+    /// Explicitly specified value by the user. Retrieved by deserializing a
+    /// non-`null` value. Can replace every other variant.
+    Specified(T),
+    /// Value that's inferred by the server. Can't replace a `Specified` variant.
+    Inferred(T),
+    /// Marker value that's retrieved when deserializing a user-specified `null`
+    /// value. Can't be used alone and has to be replaced by server-`Inferred`
+    /// or user-`Specified` value.
+    None
+}
+
+/// Deserialize as if it's `Option<T>` and use `None` variant if it's `None`,
+/// otherwise use `Specified` variant for deserialized value.
+impl<'de, T: Deserialize<'de>> Deserialize<'de> for Inferrable<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Inferrable<T>, D::Error>
+        where D: Deserializer<'de>
+    {
+        let value = Option::<T>::deserialize(deserializer)?;
+        Ok(match value {
+            None => Inferrable::None,
+            Some(value) => Inferrable::Specified(value),
+        })
+    }
+}
+
+impl<T: Clone + Debug> Inferrable<T> {
+    pub fn combine(&self, new: &Self) -> Self {
+        let new = match (self, new) {
+            // Don't allow to update a Specified value with an Inferred one
+            (&Inferrable::Specified(_), &Inferrable::Inferred(_)) |
+            // Nor allow to update with a `None` variant - it's only used to
+            // allow to deserialize a non-`Specified` variant
+            (_, &Inferrable::None) => self,
+            _ => new,
+        };
+        new.clone()
+    }
+
+    pub fn infer(&mut self, value: T) {
+        if let &mut Inferrable::Specified(_) = self {
+            trace!("Trying to infer {:?} on a {:?}", value, self);
+            return;
+        }
+
+        *self = Inferrable::Inferred(value);
+    }
+}
+
+impl<T> AsRef<T> for Inferrable<T> {
+    fn as_ref(&self) -> &T {
+        match self {
+            &Inferrable::Inferred(ref value) |
+            &Inferrable::Specified(ref value) => value,
+            &Inferrable::None => unreachable!(),
```

Comment:
In fact, I wonder that we might hit this in real life - either because inference has not completed yet, or the user resets to null. In which case we should probably be able to continue without crashing - either just skip the build or build with some defaults or something